lib/3scale/backend/oauth/token_storage.rb

```diff
@@ -185,10 +185,7 @@ def remove_whole_token_set(token_set, service_id)
             # TODO: provide a SSCAN interface with lazy enums because SMEMBERS
             # is prone to DoSing and timeouts
             def tokens_from(token_set)
-              # It is important that we make this a lazy enumerator. The
-              # laziness is maintained until some enumerator forces execution or
-              # the caller calls 'to_a' or 'force', whichever happens first.
-              storage.smembers(token_set).lazy
+              storage.smembers(token_set)
```

Comment:
I understand how `.lazy` works and I made a conscious decision when I decided to remove it. Let me try to justify that. I admit that I didn't do a good job explaining this in the description of the commit.

First of all, this code is deprecated. It's only used in the endpoints that we offer to manage OAuth tokens which are deprecated.

I spent some time trying to debug the issue, but to be honest, I was not able to fully understand why `.lazy` does not play well with the async library. It might be related to something like this: https://github.com/socketry/async/issues/23, but I do not know for sure.

Regarding your points about external/internal fragmentation:
- External fragmentation. As you explained, `.lazy()` would be absolutely critical if we changed this to SSCAN as the TODO in the code mentions, but I suspect this is not going to happen. It has not happened in years, so it's not going to happen now that the code is deprecated.

- Internal fragmentation. As far as I know, the method where the call to `.lazy` was is only called from `OAuth::Token::Storage.all_by_service_and_app` which needs to return all the tokens returned by the `smembers` call except for the expired ones. So there's a `.select` that discards some elements, but there's nothing like a `take(a_few)`.

Reply:
@davidor re: SSCAN happening - well, it can still happen and the fact this is "deprecated" does not mean `SSCAN` would be used elsewhere. The fact this code implies `lazy` won't be usable is something that will need to be documented and investigated eventually.

Reply:
I know. I added a note in the TODO list above to document this in the future. We might find other limitations in future PRs so I'll document all of them later.